initrd/etc/functions

```diff
@@ -3,26 +3,68 @@
 
 die() {
 	echo >&2 "$*";
+	# In order to die with dignity...
+	echo s > /proc/sysrq-trigger
+	echo u > /proc/sysrq-trigger
 	exit 1;
 }
 
 warn() {
 	echo >&2 "$*";
 }
 
+gpg_auth() {
+	echo >&2 "!!!!! Please authenticate with OpenPGP card to prove you own this box..."
+
+	confirm_gpg_card
+
+	# Perform a signing-based challenge-response,
+	# to authencate that the card plugged in holding
+	# the key to sign the list of boot files.
+	CR_NONCE="/tmp/secret/cr_nonce"
+	CR_SIG="/tmp/secret/cr_nonce.sig"
+
+	dd \
+		if=/dev/urandom \
+		of="$CR_NONCE" \
+		count=1 \
+		bs=20 \
+		2>/dev/null \
+		|| die "Unable to generate 20 random bytes"
+
+	for tries in 1 2 3; do
+		if gpg --digest-algo SHA256 \
+			--detach-sign \
+			-o "$CR_SIG" \
+			"$CR_NONCE" \
+			&& gpgv "$CR_SIG" "$CR_NONCE" \
+		; then
+			return 0
+		else
+			rm "$CR_SIG"
+			continue
+		fi
+	done
+	return 1
+}
+
 recovery() {
 	echo >&2 "!!!!! $*"
 
-	# Remove any temporary secret files that might be hanging around
-	# but recreate the directory so that new tools can use it.
-	rm -rf /tmp/secret
-	mkdir -p /tmp/secret
-	if [ "$CONFIG_TPM" = y ]; then
-		tpm extend -ix 4 -ic recovery
+	if [ "$CONFIG_RECOVERY_NEED_AUTH" != y ] || gpg_auth; then
+		# Remove any temporary secret files that might be hanging around
+		# but recreate the directory so that new tools can use it.
+		rm -rf /tmp/secret
```

Comment:
the `rm -rf /tmp/secret` and `mkdir -p /tmp/secret` should both be done always before the if block, right?

Reply:
It is just changed as you pointed out.